units/event_log_tests.cc

```diff
@@ -0,0 +1,183 @@
+
+  AttributeImpl(const char *_name, T _value)
+      : Attribute{_name, VT}, value(_value) {}
+  T &GetValue() { return value; }
+};
+
+using VectorInt64Attr =
+    AttributeImpl<std::vector<int64_t>, AttributeType::VectorInt64>;
+using StringAttr = AttributeImpl<std::string, AttributeType::String>;
+using Int64Attr = AttributeImpl<int64_t, AttributeType::Int64>;
+
+// Event represents the base struct for a loggable event. It contains the
+// event's name and the level of importance. The derived structs must define and
+// initialize their own set of attributes.
+struct Event {
+  const char *name;
+  LogLevel log_level;
+
+  Event(const char *_name, LogLevel _log_level)
+      : name(_name), log_level(_log_level) {}
+
+  // Each implementation of an Event contains attribute(s) and
+  // overrides this function to return them.
+  virtual const std::vector<Attribute *> &GetAttributes() = 0;
+  virtual size_t GetNumAttributes() = 0;
+
+  const char *GetEventName() { return name; }
+  LogLevel GetLogLevel() { return log_level; }
+};
+
+// An event for the CreateShaderModule function.
+// Example:
+//    const int64_t create_time_ns = create_end - create_start;
+//    Int64Attr timestamp("timestamp", 1234);
+//    Int64Attr shader_hash("hash", res.shader_hash);
+//    std::vector<Attribute *> attributes = {&shader_hash, &timestamp};
+//    CreateShaderModuleEvent("create_shader_module_ns", attributes,
+//    LogLevel::High);
+struct CreateShaderModuleEvent : Event {
+  std::vector<Attribute *> attributes;
+
+  CreateShaderModuleEvent(const char *_name,
+                          std::vector<Attribute *> &_attributes,
+                          LogLevel _log_level)
+      : Event(_name, _log_level), attributes(_attributes) {}
+  const std::vector<Attribute *> &GetAttributes() override {
+    return attributes;
+  }
+  size_t GetNumAttributes() override { return attributes.size(); }
+};
+
+// Takes an `Event` instance as an input and generates a csv string containing
+// `event`'s name and attribute values.
+std::string EventToCSVString(Event &event) {
+  const std::vector<Attribute *> &attributes = event.GetAttributes();
+
+  std::string csv_str = event.GetEventName();
+  csv_str += ",";
+  size_t e;
+  for (size_t i = 0, e = attributes.size(); i != e; ++i) {
+    switch (attributes[i]->value_type) {
+      case AttributeType::Int64: {
+        csv_str +=
+            ValueToCSVString(attributes[i]->cast<Int64Attr>()->GetValue());
+        break;
+      }
+      case AttributeType::String: {
+        csv_str +=
+            ValueToCSVString(attributes[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case AttributeType::VectorInt64: {
+        csv_str += ValueToCSVString(
+            attributes[i]->cast<VectorInt64Attr>()->GetValue());
+        break;
+      }
+    }
+    if (i + 1 != e) csv_str += ",";
+  }
+  return csv_str;
+}
+
+TEST(Event, AttributeCreation) {
+  Int64Attr timestamp("timestamp", 1601314733266018318);
```

Comment:
In this basic test we could check that the value and name can be retrieved and that they match what you passed to create each of the attributes